server/src/test/java/org/elasticsearch/snapshots/mockstore/MockEventuallyConsistentRepository.java

```diff
@@ -0,0 +1,288 @@
+            this(operation, path, null);
+        }
+    }
+
+    private class MockBlobStore implements BlobStore {
+
+        private AtomicBoolean closed = new AtomicBoolean(false);
+
+        @Override
+        public BlobContainer blobContainer(BlobPath path) {
+            return new MockBlobContainer(path);
+        }
+
+        @Override
+        public void close() {
+            closed.set(true);
+        }
+
+        private void ensureNotClosed() {
+            if (closed.get()) {
+                throw new AssertionError("Blobstore is closed already");
+            }
+        }
+
+        private class MockBlobContainer implements BlobContainer {
+
+            private final BlobPath path;
+
+            MockBlobContainer(BlobPath path) {
+                this.path = path;
+            }
+
+            @Override
+            public BlobPath path() {
+                return path;
+            }
+
+            @Override
+            public boolean blobExists(String blobName) {
+                ensureNotClosed();
+                try {
+                    readBlob(blobName);
+                    return true;
+                } catch (NoSuchFileException e) {
+                    return false;
+                }
+            }
+
+            @Override
+            public InputStream readBlob(String name) throws NoSuchFileException {
+                ensureNotClosed();
+                final String blobPath = path.buildAsString() + name;
+                synchronized (context.actions) {
+                    final List<BlobStoreAction> relevantActions = new ArrayList<>(
+                            context.actions.stream().filter(action -> blobPath.equals(action.path)).collect(Collectors.toList()));
+                    context.actions.add(new BlobStoreAction(Operation.GET, blobPath));
+                    for (int i = relevantActions.size() - 1; i > 0; i--) {
+                        if (relevantActions.get(i).operation == Operation.GET) {
+                            relevantActions.remove(i);
+                        } else {
+                            break;
+                        }
+                    }
+                    final List<byte[]> writes = new ArrayList<>();
+                    boolean readBeforeWrite = false;
+                    for (BlobStoreAction relevantAction : relevantActions) {
+                        if (relevantAction.operation == Operation.PUT) {
+                            writes.add(relevantAction.data);
+                        }
+                        if (writes.isEmpty() && relevantAction.operation == Operation.GET) {
+                            readBeforeWrite = true;
+                        }
+                    }
+                    if (writes.isEmpty()) {
+                        throw new NoSuchFileException(blobPath);
+                    }
+                    if (readBeforeWrite == false && relevantActions.size() == 1) {
```

Comment:
This sounds like a complicated way of just saying the first relevant action is a PUT?
Given that you're already filtering trailing GETs from relevantActions. Why not just check here if `relevantActions.size() == 1 && relevantActions.get(0).operation == PUT`?

Reply:
True, sorry for the noise. Simplified this now.